server/src/main/java/org/elasticsearch/action/bulk/BulkShardRequestInferenceProvider.java

```diff
@@ -0,0 +1,296 @@
+                    onBulkItemFailure.accept(
+                        bulkItemRequest,
+                        new IllegalArgumentException("No inference provider found for model ID " + modelId)
+                    );
+                    continue;
+                }
+                ActionListener<InferenceServiceResults> inferenceResultsListener = new ActionListener<>() {
+                    @Override
+                    public void onResponse(InferenceServiceResults results) {
+
+                        if (results == null) {
+                            throw new IllegalArgumentException(
+                                "No inference retrieved for model ID " + modelId + " in document " + docWriteRequest.id()
+                            );
+                        }
+
+                        int i = 0;
+                        for (InferenceResults inferenceResults : results.transformToLegacyFormat()) {
+                            String fieldName = inferenceFieldNames.get(i++);
+                            @SuppressWarnings("unchecked")
+                            List<Map<String, Object>> inferenceFieldResultList = (List<Map<String, Object>>) rootInferenceFieldMap
+                                .computeIfAbsent(fieldName, k -> new ArrayList<>());
+                            // Remove previous inference results if any
+                            inferenceFieldResultList.clear();
+
+                            // TODO Check inference result type to change subfield name
+                            var inferenceFieldMap = Map.of(
+                                SPARSE_VECTOR_SUBFIELD_NAME,
+                                inferenceResults.asMap("output").get("output"),
+                                TEXT_SUBFIELD_NAME,
+                                docMap.get(fieldName)
+                            );
+                            inferenceFieldResultList.add(inferenceFieldMap);
+                        }
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        failedItems.add(itemIndex);
+                        onBulkItemFailure.accept(bulkItemRequest, e);
+                    }
+                };
+                inferenceProvider.service()
+                    .infer(
+                        inferenceProvider.model,
+                        inferenceFieldNames.stream().map(docMap::get).map(String::valueOf).collect(Collectors.toList()),
+                        // TODO check for additional settings needed
+                        Map.of(),
+                        InputType.INGEST,
+                        ActionListener.releaseAfter(inferenceResultsListener, docRef.acquire())
+                    );
+            }
+        }
+    }
+
+    private static List<String> getFieldNamesForInference(Map.Entry<String, Set<String>> fieldModelsEntrySet, Map<String, Object> docMap) {
+        List<String> inferenceFieldNames = new ArrayList<>();
```

Comment:
Minor optimization: We could pre-allocate an `ArrayList` of the maximum required size by using the inference field set size

Reply:
I tend not to do that for small lists - it will be expanded to 10 elements when the first element is added, so probably just removes one list expansion.